src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/VsUtilities.cs

```diff
@@ -0,0 +1,182 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.TextManager.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor
+{
+    internal static class VsUtilities
+    {
+        public static void ApplyTextEdits(
+            IServiceProvider serviceProvider,
+            Uri uri,
+            ITextSnapshot snapshot,
+            IEnumerable<TextEdit> textEdits)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            ApplyTextEdits(textEdits, snapshot, snapshot.TextBuffer);
+
+            var cursorPosition = ExtractCursorPlaceholder(snapshot.TextBuffer.CurrentSnapshot);
+            if (cursorPosition != null)
+            {
+                var fullPath = GetLocalFilePath(uri);
+
+                VsShellUtilities.OpenDocument(serviceProvider, fullPath, VSConstants.LOGVIEWID.TextView_guid, out _, out _, out var windowFrame);
+
+                if (windowFrame != null)
+                {
+                    var textView = GetActiveVsTextView(windowFrame);
+                    MoveCaretToPosition(textView, cursorPosition);
+                }
+            }
+        }
+
+        // Internal for testing
+        internal static Position ExtractCursorPlaceholder(ITextSnapshot snapshot)
+        {
+            Position cursorPosition = null;
+            var text = snapshot.GetText();
```

Comment:
This is creating a string of the whole document text. Bad. But I found this to be the least complicated way to do this. Otherwise, finding out the line, column of the placeholder gets really complicated.